datafusion/src/physical_plan/file_format/parquet.rs

```diff
@@ -385,9 +388,33 @@ fn build_row_group_predicate(
     }
 }
 
+// Map projections from the schema which merges all file schemas to projections on a particular
+// file
+fn map_projections(
```

Comment:
I wonder if this logic and the logic in read_partition might be extracted into some of RecordBatchAdapter, akin to `PartitionColumnProjector`. This would allow the logic to be reused with other file formats, e.g. JSON or CSV, whilst also allowing testing it in isolation.

Reply:
that is a good idea. If not in this PR I'll file a ticket to do it in a follow on

Reply:
https://github.com/apache/arrow-datafusion/issues/1669